Add pkgsCross.emscripten, replacing hand-crafted emscriptenPackages

AtnNn opened the pull request:
###### Description of changes

- Added a pre-built Emscripten `SYSTEM` cache to the `emscripten` package
- Added `emscriptenFull` which additionally contains Emscripten `PORTS`
- Added `emscriptenNoCache` which requires setting `EM_CACHE` to use
- Add wrapped `emscriptenCC` and `emscriptenBintools` 
- Added Emscripten to `lib/systems` and `pkgsCross`
- Moved the `xmlmirror` package from inside `top-level/emscripten-packages.nix` to `tools/text/xml`
- Removed custom `emscriptenPackages.json_c` and `emscriptenPackages.zlib`
- Removed `buildEmscriptenPackage`
- Added some `pkgsCross.emscripten` packages to `release-cross.nix`

Possibly fixes #28929 #141246 #136396. Improves the fix for #139943.

Tested with

```
$ nix-build --show-trace -E 'with import ./pkgs/top-level/release-cross.nix {}; builtins.mapAttrs (k: v: v."x86_64-linux") emscripten'
$ node result-3/bin/hello
Hello, world!
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

AtnNn commented:
What do you think about this approach? @qknight @matthewbauer @RaitoBezarius

RaitoBezarius commented:
> What do you think about this approach? @qknight @matthewbauer @RaitoBezarius

Interesting changes, thank you for doing this. I am not too expert in the stdenv area, so I would rather defer to people working on that, cc @alyssais @Ericson2314 

I will need more time to review though and test thoroughly, but it seems to be a PR in the right direction.

alyssais commented on lib/systems/examples.nix:
```diff
@@ -331,4 +331,9 @@ rec {
+
+  # Emscripten
+  emscripten = {
+    config = "wasm32-emscripten";
```
```suggestion
    config = "wasm32-unknown-emscripten";
```

alyssais commented on lib/systems/inspect.nix:
```diff
@@ -72,6 +72,7 @@ rec {
     isCygwin       = { kernel = kernels.windows; abi = abis.cygnus; };
     isMinGW        = { kernel = kernels.windows; abi = abis.gnu; };
     isWasi         = { kernel = kernels.wasi; };
+    isEmscripten = { kernel = kernels.emscripten; };
```
Please make sure to keep alignment intact if the surrounding code is using it.

alyssais commented on pkgs/development/compilers/emscripten/wrapped.nix:
```diff
@@ -0,0 +1,84 @@
+# TODO ATN: this is missing a lot of features that cc-wrapper has, but cc-wrapper also has features that conflict with emscripten. What to do?
```
Can you explain in a bit more detail what the conflicts are?

AtnNn replied:
I'm not sure anymore. I've rewritten this PR a little to wrap Emscripten in a way that can be `wrapCC`ed, and that seems to work better.

alyssais commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -381,7 +381,16 @@ else let
+          ++ lib.optionals (stdenv.isEmscripten && stdenv ? cc.cc) [
+
+             # TODO ATN: I'm not sure where to put this, it doesn't look like it belongs here
+            "-DCMAKE_TOOLCHAIN_FILE=${stdenv.cc.cc}/share/emscripten/cmake/Modules/Platform/Emscripten.cmake"
```
Does a "native" Emscripten build make sense (running Nix on Emscripten)?  If not, put this in crossFlags above, where it'll be at home next to the Darwin one.

alyssais commented on pkgs/tools/misc/figlet/default.nix:
```diff
@@ -23,6 +23,12 @@ stdenv.mkDerivation rec {
     })
   ];
 
+  # TODO ATN: How to deal with `emcc' generating split binaries? Should `cp` and `install` be wrapped to also copy associated `.wasm` files? Should patches like this be added all over? Should `emcc` be wrapped to somehow combine the split binaries?
```
Can you explain more about these split binaries?  When do you we need each part?  Wrapping `cp` and `install` doesn't sound feasible — there are many more ways to move files around.

RaitoBezarius replied:
They are mentioning this feature: https://emscripten.org/docs/optimizing/Module-Splitting.html
emcc can generate multiple outputs : JS & WASM, it is not obvious that classical tooling would deal with the multiple outputs.

alyssais commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20075,12 +20072,14 @@ with pkgs;
+    callPackage ../development/libraries/gobject-introspection {
+      nixStoreDir = config.nix.storeDir or builtins.storeDir;
+      inherit (darwin) cctools;
+    };
```
Probably best not to reformat code you aren't otherwise touching.

alyssais reviewed (changes requested):
This is a very promising direction.  Thank you for exploring it!  Since it's still early, I've focused my review only on the stdenv stuff, not any individual package modifications.

The biggest threat I see is the split binaries, so some more info on those would be great.

AtnNn commented:
Thanks @alyssais. I'm going to update this branch to use Emscripten's `-sSINGLE_FILE` mode, which seems to solve the issues I had with split executables.

mohe2015 commented:
Could you please fix the syntax errors in overlay.nix? And maybe update to latest nixpkgs?

RaitoBezarius commented:
Let's get this after the branch-off :).

willcohen commented:
Just as a note for a parallel emscripten change, #229718 (not merged as of this comment) would bump underlying emscripten from 3.1.24 to 3.1.39, which in turn requires LLVM 16 instead of 14.

SuperSandro2000 commented on lib/strings.nix:
```diff
@@ -380,6 +380,11 @@ rec {
   */
   escapeShellArg = arg: "'${replaceStrings ["'"] ["'\\''"] (toString arg)}'";
 
+  escapeShellArg' = arg:
+    if match "^[-%+./0-9:=@A-Z_a-z]+$" arg != null
+    then arg
+    else escapeShellArg arg;
```
This must get a comment when it is useful

SuperSandro2000 commented on lib/systems/inspect.nix:
```diff
@@ -59,7 +59,7 @@ rec {
     isBSD          = { kernel = { families = { inherit (kernelFamilies) bsd; }; }; };
     isDarwin       = { kernel = { families = { inherit (kernelFamilies) darwin; }; }; };
-    isUnix         = [ isBSD isDarwin isLinux isSunOS isCygwin isRedox ];
+    isUnix         = [ isBSD isDarwin isLinux isSunOS isCygwin isRedox isEmscripten ];
```
I highly doubt this is unix

alyssais replied:
It is!

> It even emulates an entire [POSIX](https://en.wikipedia.org/wiki/POSIX) operating system, enabling programmers to use functions from the [C standard library](https://en.wikipedia.org/wiki/C_standard_library) (libc).

https://en.wikipedia.org/wiki/Emscripten

SuperSandro2000 commented on lib/systems/parse.nix:
```diff
@@ -294,6 +294,7 @@ rec {
     openbsd  = { execFormat = elf;     families = { inherit bsd; }; };
     solaris  = { execFormat = elf;     families = { }; };
     wasi     = { execFormat = wasm;    families = { }; };
+    emscripten     = { execFormat = wasm;    families = { }; };
```
```suggestion
    emscripten = { execFormat = wasm;    families = { }; };
```

alyssais replied:
Best would be to adjust the alignment of the others to match.

SuperSandro2000 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -14,7 +14,10 @@
-, isGNU ? false, isClang ? cc.isClang or false, gnugrep ? null
+, isGNU ? false
+, isClang ? cc.isClang or false
+, isEmscripten ? cc.isEmscripten or false
```
This is not used

SuperSandro2000 commented on pkgs/stdenv/emscripten/overlay.nix:
```diff
@@ -0,0 +1,38 @@
+    libogg = ogg;
+    sqlite = sqlite3;
+    libvorbis = vorbis;
+    zlib = zlib;
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/libraries/gobject-introspection/default.nix:
```diff
@@ -89,10 +89,13 @@ stdenv.mkDerivation (finalAttrs: {
-    libffi
-    glib
-  ];
+  propagatedBuildInputs =
+    if stdenv.isEmscripten
+    then []
+    else [
+      libffi
+      glib
+    ];
```
```suggestion
  propagatedBuildInputs = lib.optional (!stdenv.isEmscripten) [
    libffi
    glib
  ];
```

SuperSandro2000 commented on pkgs/development/compilers/emscripten/ports.nix:
```diff
@@ -0,0 +1,136 @@
+    sha256 = "sha256-hzXN30IocklLty8daZSF5rHQJ4xHKLWzCQP7o0m5Z0s=";
+  };
+  libpng  = {
+    url = "https://github.com/libsdl-org/SDL/archive/release-2.24.0.zip";
```
At least this URL is duplicated. Not sure if it is worth to dedupe. Also can't we re-use the src's from our packages?

AtnNn replied:
I've fixed the duplicate URLs. The src's from nixpkgs cannot be used here, the `url` in this file point to the exact sources that `embuilder.py` requires.

SuperSandro2000 commented on pkgs/development/compilers/emscripten/install-wrapper.sh:
Is this supposed to be empty?

AtnNn replied:
Unused and removed!

SuperSandro2000 commented on pkgs/development/compilers/emscripten/emscripten.nix:
```diff
@@ -0,0 +1,136 @@
+  meta = with lib; {
+    homepage = "https://github.com/emscripten-core/emscripten";
+    description = "An LLVM-to-JavaScript Compiler";
+    platforms = platforms.all;
```
Not sure if unix would be better here

AtnNn replied:
This matches `platforms.all` in the existing `emscripten` package.

SuperSandro2000 commented on pkgs/development/compilers/emscripten/emscripten.nix:
```diff
@@ -0,0 +1,136 @@
+
+    export EM_FROZEN_CACHE=1
+
+    export PYTHON=${python3}/bin/python
```
```suggestion
    export PYTHON=${python3.forBuild}/bin/python
```

AtnNn replied:
I was able to remove this line entirely, which should address the issue.

SuperSandro2000 commented on pkgs/development/compilers/emscripten/emscripten.nix:
```diff
@@ -0,0 +1,136 @@
+        --set NODE_PATH ${nodeModules} \
+        --set EM_EXCLUSIVE_CACHE_ACCESS 1 \
+        --set PYTHON ${python3}/bin/python \
+        --set PYTHONPATH '${python3Packages.makePythonPath [ python3Packages.requests ]}'
```
```suggestion
        --set PYTHONPATH '${python3.pkgs.makePythonPath [ python3.pkgs.requests ]}'
```

AtnNn replied:
Is there an advantage to this? #211340 suggests that we should

> Discourage using package sets like python3.pkgs "directly"

SuperSandro2000 commented on pkgs/development/compilers/emscripten/emscripten.nix:
```diff
@@ -0,0 +1,136 @@
+  };
+
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ nodejs python3 ];
```
If this is only used in the multi line strings below then this is no-op

SuperSandro2000 commented on pkgs/development/compilers/emscripten/emscripten.nix:
```diff
@@ -0,0 +1,136 @@
+
+    dontBuild = true;
+
+    # Copy node_modules directly.
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/compilers/emscripten/emscripten.nix:
```diff
@@ -0,0 +1,136 @@
+{ lib, stdenv, fetchFromGitHub, python3, nodejs, closurecompiler, python3Packages
```
```suggestion
{ lib, stdenv, fetchFromGitHub, python3, nodejs, closurecompiler
```
use python3.pkgs

SuperSandro2000 commented on pkgs/development/compilers/emscripten/default.nix:
```diff
@@ -1,149 +1,33 @@
-    sed -i "s|^EMRANLIB =.*|EMRANLIB='$out/bin/emranlib'|" tools/shared.py
-
-    runHook postBuild
+  unpackPhase = ''
+    src=$PWD
   '';
```
Can we use sourceRoot instead?

SuperSandro2000 commented on pkgs/development/compilers/emscripten/default.nix:
```diff
@@ -1,149 +1,33 @@
-  };
+stdenvNoCC.mkDerivation {
+  pname = "emscripten-with-cache";
+  version = emscriptenPlain.version;
```
```suggestion
  inherit (emscriptenPlain) version;
```

SuperSandro2000 commented on pkgs/development/compilers/emscripten/default.nix:
```diff
@@ -1,149 +1,33 @@
-    bintools = emscripten;
-  };
-
-  meta = with lib; {
```
We must keep a meta

SuperSandro2000 commented on pkgs/development/compilers/emscripten/cc.nix:
```diff
@@ -0,0 +1,55 @@
+  dontConfigure = true;
+  enableParallelBuilding = true;
+
+  buildInputs = [ makeWrapper ];
```
```suggestion
  nativeBuildInputs = [ makeWrapper ];
```

AtnNn replied:
Thanks! I missed your comment and wasted a few hours finding this bug.

SuperSandro2000 commented on pkgs/development/compilers/emscripten/cc.nix:
```diff
@@ -0,0 +1,55 @@
+  buildInputs = [ makeWrapper ];
+
+  unpackPhase = ''
+    src=$PWD
```
Can we use sourceRoot?

AtnNn replied:
I just copied this from `pkgs/build-support/bintools-wrapper/default.nix`. How would I use `srcRoot` instead?

sternenseemann replied:
This is okay, the point is to forgo upacking (since there is no source), but `src` needs to be set to something. `sourceRoot` only controls the working directory which is irrelevant here.

SuperSandro2000 commented on pkgs/development/compilers/emscripten/cc.nix:
```diff
@@ -0,0 +1,55 @@
+
+  preferLocalBuild = true;
+
+  passthru = {
```
passthru should be before meta

SuperSandro2000 commented on pkgs/development/compilers/emscripten/cache.nix:
```diff
@@ -0,0 +1,64 @@
+    export EM_CACHE=$out/share/emscripten/cache
+    mkdir -p $EM_CACHE
+
+  '' + lib.concatStrings (builtins.map (base: ''
```
```suggestion
  '' + lib.concatStrings (map (base: ''
```

SuperSandro2000 commented on pkgs/development/compilers/emscripten/cache.nix:
```diff
@@ -0,0 +1,64 @@
+asLibrary = emscriptenStdenv:
+  assert emscriptenStdenv.hostPlatform.isEmscripten;
+  emscriptenStdenv.mkDerivation {
+    pname = "${if names == "" then "unknown" else names}";
```
```suggestion
    pname = if names == "" then "unknown" else names;
```

SuperSandro2000 commented on pkgs/development/compilers/emscripten/bintools.nix:
```diff
@@ -0,0 +1,63 @@
+  dontConfigure = true;
+  enableParallelBuilding = true;
+
+  buildInputs = [ makeWrapper ];
```
```suggestion
  nativeBuildInputs = [ makeWrapper ];
```

alyssais commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -390,7 +390,9 @@ else let
-          explicitFlags ++ lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform) crossFlags;
+          explicitFlags
+          ++ lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform) crossFlags
+          ++ (stdenv.extraCmakeFlags or []) ;
```
The parens aren't necessary either, I think.

mohe2015 replied:
I would prefer keeping the parens as it makes it more clear what is meant.

SuperSandro2000 replied:
I would trust deadnix. If it says they are optional, I would remove them.

willcohen commented:
Noting that #229718, which bumps underlying emscripten and llvm, has entered staging.

Artturin commented on lib/systems/default.nix:
```diff
@@ -191,24 +194,33 @@ rec {
         emulatorAvailable = pkgs: (selectEmulator pkgs) != null;
 
-        emulator = pkgs:
+        emulatorNeeded = pkgs:
```
Unused and there's already canExecute

AtnNn commented:
Thanks for the feedback. I've cleaned up the code and updated the original implementation:

- Use `-s SINGLE_FILE`, which allows `figlet` and `hello` to build working binaries.
- Use `wrapCC` and `wrapBintools` 
- Test with more packages such as `gmp` and `sqlite`
- Support Emscripten Ports

Artturin commented on lib/strings.nix:
```diff
@@ -425,18 +425,41 @@ rec {
 
+  /* Like escapeShellArg, but doesn't add redundant quotes.
+
+     Used to avoid mass rebuilds when replacing e.g. `"cp -r"` with
+     `escapeShellArgs [ "cp" "-r" ]`.
```
very bad, this function shouldn't be added. just target staging

Artturin commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1147,7 +1147,10 @@ with pkgs;
   makeInitrdNGTool = callPackage ../build-support/kernel/make-initrd-ng-tool.nix { };
 
-  makeWrapper = makeShellWrapper;
+  makeWrapper =
+    if stdenv.hostPlatform.isWindows || stdenv.hostPlatform.isEmscripten
+    then null
```
should throw or something

comment why it doesn't work on emscripten

can't be changed to makeBinaryWrapper on windows https://github.com/NixOS/nixpkgs/issues/120726#issuecomment-1518984250

AtnNn replied:
My intent here is to allow packages to conditionally use `makeWrapper` if it is available by checking for `makeWrapper == null`. Changing this to a throw wouldn't work for that. Is there a better way to do that? Maybe something like `lib.elem stdenv.hostPlatform makeWrapper.meta.platforms`?

sternenseemann replied:
Indeed, the using expression should check whether something is available or not. `lib.meta.availableOn` should suffice for this, given that the compatibility is expressed in (`makeWrapper`'s) `meta.platforms`.

Using `null` will also break `makeWrapper` in `buildPackages` for the affected package sets due to splicing depending on attribute sets.

Artturin commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -380,7 +380,7 @@ else let
         ] ++ lib.optionals (stdenv.buildPlatform.uname.release != null) [
           "-DCMAKE_HOST_SYSTEM_VERSION=${stdenv.buildPlatform.uname.release}"
-        ]);
+        ]) ++ (stdenv.extraCmakeFlags or []);
```
there's already a cmakeFlags ++ above

sternenseemann replied:
I think this is better achieved with a setup hook that appends to `cmakeFlags`.

Artturin commented on pkgs/stdenv/cross/default.nix:
```diff
@@ -83,8 +90,13 @@ in lib.init bootStages ++ [
              buildPackages.updateAutotoolsGnuConfigScriptsHook
-        ;
+          ;
+      extraAttrs = {
```
extraAttrs here only effects the stdenv build

sternenseemann commented on pkgs/development/compilers/emscripten/bintools.nix:
```diff
@@ -0,0 +1,66 @@
+  passthru = {
+    libc_bin = "";
+    libc_dev = "";
+    libc_lib = "";
```
`null` is better here.

sternenseemann commented on pkgs/top-level/all-packages.nix:
```diff
@@ -7469,10 +7472,6 @@ with pkgs;
 
-  ### DEVELOPMENT / EMSCRIPTEN
-
-  buildEmscriptenPackage = callPackage ../development/em-modules/generic { };
```
This is a substantial, breaking change and should also be explained in the release notes for 23.11.

sternenseemann commented on doc/languages-frameworks/emscripten.section.md:
```diff
@@ -4,179 +4,51 @@
 
-* nix
-* nixpkgs
+This mode is far more power full since this makes use of `nix` for
```
```suggestion
This mode is far more powerful since this makes use of `nix` for
```

Though this sentence doesn't really make sense: More powerful than what? Imperative Usage is only described later on…

sternenseemann commented on lib/systems/default.nix:
```diff
@@ -212,24 +215,31 @@ rec {
         emulatorAvailable = pkgs: (selectEmulator pkgs) != null;
 
-        emulator = pkgs:
+        emulatorCommand = pkgs:
```
I dislike introducing yet another field here, how is this necessary?

sternenseemann commented on pkgs/development/libraries/libxml2/default.nix:
```diff
@@ -27,7 +27,9 @@ let
   # Beware that the old version has known security issues, so the final set shouldn't use it.
-  oldVer = python.pname == "python3-minimal";
+  oldVer =
+    (pythonSupport || stdenv.isDarwin) # testing for isDarwin only to avoid mass rebuild (stdenv contains libxml2)
```
Just target staging.

sternenseemann commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -172,7 +174,7 @@ let
       mv $out/bin $bin/bin
 
-    '' + lib.optionalString (!stdenv.hostPlatform.isWindows)
+    '' + lib.optionalString (makeWrapper != null)
```
`lib.meta.availableOn makeWrapper stdenv.hostPlatform` would be a possibility or querying `meta.broken` of `makeWrapper`—or just expanding the conditional.

For better or worse, the convention in nixpkgs is that the expression using a package decides whether it can use it or not. This also has the sense that it prevents stuff from being silently unavailable which can alter configure time settings etc.

AtnNn commented:
Thanks for the suggestions. I've:
- removed `escapeShellArgs'` and `emulatorCommand`
- used a setup hook for CMake flags
- stopped avoiding rebuilds and target staging instead

Artturin commented on pkgs/build-support/setup-hooks/emscripten-cmake-hook.sh:
```diff
@@ -0,0 +1,9 @@
+    )
+}
+
+addEnvHooks "$hostOffset" addEmscriptenCMakeEnv
+
+
```
trailing newlines should be removed

Artturin commented on pkgs/development/compilers/emscripten/bintools.nix:
```diff
@@ -0,0 +1,66 @@
+  targetPrefix = "${wasmArch}-emscripten-";
+  llvm = emscripten.llvmEnv;
+
+unwrapped = stdenv.mkDerivation {
```
wrong indentation,  run `git diff --name-only HEAD~ --diff-filter=A | grep "\.nix" | xargs nixpkgs-fmt` to format added files only

Artturin commented on pkgs/development/compilers/emscripten/bintools.nix:
```diff
@@ -0,0 +1,66 @@
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  unpackPhase = ''
+    src=$PWD
+  '';
```
```suggestion
  dontUnpack = true;
```
same for the others

Artturin commented on pkgs/development/compilers/emscripten/bintools.nix:
```diff
@@ -0,0 +1,66 @@
+    src=$PWD
+  '';
+
+  installPhase = ''
```
`runHook`'s missing
also in other files

Artturin commented on pkgs/development/compilers/emscripten/bintools.nix:
```diff
@@ -0,0 +1,66 @@
+    lnLLVM llvm-cxxfilt c++filt
+    lnLLVM llvm-addr2line addr2line
+
+    echo -n > $out/nix-support/libc-ldflags-before
```
why?, add a comment.

Artturin commented on pkgs/top-level/all-packages.nix:
```diff
@@ -21135,6 +21138,7 @@ with pkgs;
     else if name == "nblibc" then targetPackages.netbsdCross.libc or netbsdCross.libc
     else if name == "wasilibc" then targetPackages.wasilibc or wasilibc
     else if name == "relibc" then targetPackages.relibc or relibc
+    else if stdenv.targetPlatform.isEmscripten then null
```
why is emscripten special from all the others, why is `stdenv.targetPlatform.libc` not null on emscripten

AtnNn replied:
I'll remove this and set `libc = null`

Artturin commented on pkgs/shells/bash/5.nix:
```diff
@@ -136,7 +136,7 @@ stdenv.mkDerivation rec {
     '';
     license = licenses.gpl3Plus;
     platforms = platforms.all;
-    maintainers = with maintainers; [ dtzWill ];
```
accidentally removed

Artturin commented on lib/tests/systems.nix:
```diff
@@ -95,6 +95,7 @@ lib.runTests (
         # The functions should be derived from the data, so this is not a problem.
         canExecute = null;
         emulator = null;
+        emulatorCommand = null;
```
```suggestion
```
leftover

Artturin commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -125,3 +125,5 @@ The module update takes care of the new config syntax and the data itself (user
   can automatically format the root device by setting
   `virtualisation.fileSystems."/".autoFormat = true;`.
+
+- `emscriptenPackages` and `buildEmscriptenPackage` have been replaced with `pkgsCross.emscripten` and `pkgsCross.emscripten.stdenv.mkDerivation`.
```
don't add directly at the end but at a random location to prevent merge conflicts

AtnNn committed:
emscripten: replace emscriptenPackages with pkgsCross.emscripten

willcohen commented:
As a quick note I've discovered working through https://github.com/NixOS/nixpkgs/pull/260267/ -- emscripten will now need access to emscripten's node_modules folder as well, so it can run `acorn-optimizer.js` during some of its passes.

allsey87 commented:
I was experimenting a bit with this PR and considering to revive it and noticed that the following works:
```
nix-build '<nixpkgs>' -I nixpkgs=https://github.com/AtnNn/nixpkgs/archive/emscripten.tar.gz \
  -A pkgsCross.emscripten.hello
```
However, this does not work:
```
nix-build '<nixpkgs>' -I nixpkgs=https://github.com/AtnNn/nixpkgs/archive/emscripten.tar.gz \
  --arg crossSystem '{ config = "wasm32-unknown-emscripten"; }' \
  -A hello
```
Is this expected?
